src/OptimalControlUtilities/include/BipedalLocomotionControllers/OptimalControlUtilities/OptimizationProblemElements.h

```diff
@@ -0,0 +1,254 @@
+                             const iDynTree::IndexRange& indexRange);
+    };
+
+    std::vector<EqualityConstraint> m_equalityConstrains; /**< Vector containing all the equality
+                                                             constraints */
+    std::vector<InequalityConstraint> m_inequalityConstrains; /**< Vector containing all the
+                                                                 inequality constraints */
+
+    iDynTree::VectorDynSize m_lowerBound; /**< Lower bound */
+    iDynTree::VectorDynSize m_upperBound; /**< Upper bound */
+    iDynTree::MatrixDynSize m_constraintMatrix; /**<  Constraint matrix */
+
+    int m_numberOfVariables; /**< Number of variables */
+    int m_numberOfConstraints{0}; /**< Number of constraint */
+
+    /**
+     * Get the index of the next constraint
+     * @return the index of the next constraint
+     */
+    int getNextConstraintIndex() const;
+
+public:
+    /**
+     * Bounds wraps an std::pair<iDynTree::VectorDynSize&, iDynTree::VectorDynSize&>.
+     * It contains the lower and the upper bounds
+     */
+    class Bounds : public std::pair<iDynTree::VectorDynSize&, iDynTree::VectorDynSize&>
+    {
+        // obscure first and second attribute
+        using std::pair<iDynTree::VectorDynSize&, iDynTree::VectorDynSize&>::first;
+        using std::pair<iDynTree::VectorDynSize&, iDynTree::VectorDynSize&>::second;
+
+    public:
+        // use the std::pair constructors
+        using std::pair<iDynTree::VectorDynSize&, iDynTree::VectorDynSize&>::pair;
+
+        iDynTree::VectorDynSize& lowerBound();
+        const iDynTree::VectorDynSize& lowerBound() const;
+        iDynTree::VectorDynSize& upperBound();
+        const iDynTree::VectorDynSize& upperBound() const;
+    };
+
+    /**
+     * Constructor
+     * @param handler variable handler object
+     */
+    Constraints(const VariableHandler& handler);
+
+    /**
+     * Add a new constraint. The constraint matrix and the bound vector are resized
+     * @param element pointer to the constraint
+     */
+    void addConstraint(ControlTask* element);
+
+    /**
+     * Add a new constraint. The constraint matrix and the bound vector are resized
+     * @param element pointer to the constraint
+     */
+    void addConstraint(InequalityConstraintElement* element);
+
+    /**
+     * Get (and compute) the lower and the upper bounds
+     * @return the lower and upper bounds
+     */
+    Bounds getBounds();
+
+    /**
+     * Get (and compute) the constraint matrix
+     * @return the constraint matrix
+     */
+    const iDynTree::MatrixDynSize& getConstraintMatrix();
+
+    /**
+     * Get the number of constraints
+     * @return the number of constraints
+     */
+    int getNumberOfConstraints() const;
+
+    /**
+     * Get the equality constraints
+     * @return a vector containing the reference to the equality constraints
+     */
+    const std::vector<EqualityConstraint>& getEqualityConstraints() const;
```

Comment:
In view of using smart pointers, I would avoid this method. The risk is having cyclic references (https://www.learncpp.com/cpp-tutorial/15-7-circular-dependency-issues-with-stdshared_ptr-and-stdweak_ptr/). Since the user is passing you a pointer to the object, you can assume that he can access it also afterwards.